### PR TITLE
drivers: sensors: Add lsm303dlhc driver

### PR DESCRIPTION
**Pull request by aakkds**
This sensor has a different I2C address to access accelerometer/magnetometer values, so I made two drivers.
Unfortunately I didn't add the interrupt based control as the board I used (flora from Adafruit) doesn't provide access to INT pads.

Tested on a nucleo-f446 board : 
- Accelerometer only
- Magnetometer only
- Accelerometer and magnetometer together (not tested accelerometer on one I2C bus and magnetometer on another)

### CODE REVIEW

**Review by lpereira — APPROVED** [review]
LGTM, althought I'm not familiar with these particular devices.

File: drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c
Comment (MaureenHelm) [inline drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c]: This should be static

File: drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c
Comment (MaureenHelm) [inline drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c]: This should be static

File: drivers/sensor/lsm303dlhc_accel/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/lsm303dlhc_accel/Kconfig]: Would it be simpler to replace all these bool configs with one int config, and limit the range?

File: drivers/sensor/lsm303dlhc_magneto/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/lsm303dlhc_magneto/Kconfig]: Same comment as before about replacing all the bools with one int config

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (MaureenHelm) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: static

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (MaureenHelm) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: static

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by MaureenHelm** [comment]
Please also add this sensor to tests/drivers/build_all/sensors_a_m.conf so it gets included in build testing

File: drivers/sensor/lsm303dlhc_accel/Kconfig
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_accel/Kconfig]: It would be simpler yes. But by doing so, nothing prevents the user from entering an unsupported value. In this case, it will either result in a compilation error as ODR macro won't be defined or the driver using a default value. So should I use a default value (compilation always ok but may not necessarily correspond to what the user expected) or no default value (compilation only ok if correct value entered) ?

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (ydamigos) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: output?

**Review by ydamigos — COMMENTED** [review]
I tested the drive on stm32f3_disco (PR #6441). The magnetic sensor always returns the same value on stm32f3_disco.

**Comment by galak** [comment]
> I tested the drive on stm32f3_disco (PR #6441). The magnetic sensor always returns the same value on stm32f3_disco.

Are you expecting that to be the case?  Wasn't clear if that is a bug or correct behavior :).

**Comment by ydamigos** [comment]
>Are you expecting that to be the case? Wasn't clear if that is a bug or correct behavior :).

I was expecting the magnetic sensor's measurement to change when I put a magnet next to it.

**Comment by aakkds** [comment]
@galak It's a bug, I need to enable continuous update. I think I did it during testing and then removed it :( I m looking into it now.

**Comment by aakkds** [comment]
@ydamigos The magn sensor should now be updated. That would be great if you could test it on your board :) Thanks again for testing this!

File: drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c
Comment (ydamigos) [inline drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c]: SYS_LOG_ERR?

**Review by ydamigos — CHANGES_REQUESTED** [review]
> @ydamigos The magn sensor should now be updated. That would be great if you could test it on your board :) Thanks again for testing this!

Thanks for the changes. It works now.

Please change all SYS_LOG_DBG with SYS_LOG_ERR if they are referring to an error condition.

File: drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c]: Yes, will do tonight (UTC+1)

**Comment by ydamigos** [comment]
Please also check if you add the following block in tests/drivers/build_all/dts.fixup, will fix the shippable errors:

```
#ifndef CONFIG_LSM303DLHC_ACCEL_NAME
#define CONFIG_LSM303DLHC_ACCEL_NAME ""
#define CONFIG_LSM303DLHC_ACCEL_I2C_MASTER_DEV ""
#define CONFIG_LSM303DLHC_ACCEL_I2C_ADDR 0x19
#endif

#ifndef CONFIG_LSM303DLHC_MAGN_NAME
#define CONFIG_LSM303DLHC_MAGN_NAME ""
#define CONFIG_LSM303DLHC_MAGN_I2C_MASTER_DEV ""
#define CONFIG_LSM303DLHC_MAGN_I2C_ADDR 0x1e
#endif
```

File: drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.c]: done

**Comment by aakkds** [comment]
@ydamigos Thanks for the dts.fixup suggestion! I added it

**Comment by ydamigos** [comment]
>Thanks for the changes. It works now.

@aakkds Unfortunately, I still have problems with the magnetic sensor measurement. After reset, sometimes it works, some it doesn't (the value doesn't update).

**Comment by aakkds** [comment]
@ydamigos mhh I can't reproduce. I'll try from your branch. When it doesn't update, it's the 3 values which doesn't change, event with the magnet?

**Comment by ydamigos** [comment]
>When it doesn't update, it's the 3 values which doesn't change, event with the magnet?

Yes, I posted part of the console output below.

```
***** BOOTING ZEPHYR OS v1.10.99 - BUILD: Mar  6 2018 18:34:49 *****                                             
Magnetometer data:                                                                                                
( x y z ) = ( 0.155454  -0.530000  0.355454 )                                                                     
Accelerometer data:                                                                                               
( x y z ) = ( 0.285164  0.875026  -0.507827 )                                                                     
Magnetometer data:                                                                                                
( x y z ) = ( 0.155454  -0.530000  0.355454 )                                                                     
Accelerometer data:                                                                                               
( x y z ) = ( 0.277352  0.839869  -0.500015 )                                                                     
Magnetometer data:                                                                                                
( x y z ) = ( 0.155454  -0.530000  0.355454 )                                                                     
Accelerometer data:                                                                                               
( x y z ) = ( 0.281258  0.843775  -0.488296 )                                                                     
Magnetometer data:                                                                                                
( x y z ) = ( 0.155454  -0.530000  0.355454 )                                                                     
Accelerometer data:                                                                                               
( x y z ) = ( 0.070314  0.964873  -0.343760 )                                                                     
Magnetometer data:                                                                                                
( x y z ) = ( 0.155454  -0.530000  0.355454 )                                                                     
Accelerometer data:                                                                                               
( x y z ) = ( 0.062501  0.964873  -0.324228 )                                                                     
Magnetometer data:                                                                                                
( x y z ) = ( 0.155454  -0.530000  0.355454 )
```

**Comment by aakkds** [comment]
@ydamigos I tested with your sample (with printk instead of printf, without sensor_value_to_double, as I didn't build with newlib) and did ~50 resets and could not reproduce :(

**Comment by ydamigos** [comment]
>@ydamigos I tested with your sample (with printk instead of printf, without sensor_value_to_double, as I didn't build with newlib) and did ~50 resets and could not reproduce :(

>Edit: I found this https://electronics.stackexchange.com/questions/101753/stm32f3discovery-lsm303dlhc-magnetic-field-data-not-updating
If it's really coming from the sensor (I didn't find other sources, so I doubt it), I'll remove continuous update and add the one shot mode but I don't get the fact that I can't reproduce...

I'll try to find sometime time and connect a logic analyzer to check it.

**Comment by ydamigos** [comment]
>I'll try to find sometime time and connect a logic analyzer to check it.

When the magnetic sensor measurement doesn't update, the status register always has the value 3 (data ready bit and lock bit are set).

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h
Comment (ydamigos) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h]: Data rate output bits are 3 with value between 0 to 7. LSM303DLHC_MAGN_ODR_IDX should begin from 0 and end to 7.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h
Comment (ydamigos) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h]: This should be 1.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h
Comment (ydamigos) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h]: This should be 3.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h
Comment (ydamigos) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.h]: This should be 2.

**Review by ydamigos — CHANGES_REQUESTED** [review]

**Comment by aakkds** [comment]
@ydamigos Sorry for the late reply, I addressed your changes.
Concerning the data register, when I monitor it, it never gets stucked, it just oscillates between 1 and 3 depending if I try to sample it faster than the output data rate.
I can change the magneto driver to use it in one shot mode, I guess this should work on your board but it doesn't sound like the right think to do. 
Would that work for you, or you have another suggestion?

**Comment by ydamigos** [comment]
>Concerning the data register, when I monitor it, it never gets stucked, it just oscillates between 1 and 3 depending if I try to sample it faster than the output data rate.
I can change the magneto driver to use it in one shot mode, I guess this should work on your board but it doesn't sound like the right think to do.
Would that work for you, or you have another suggestion?

@aakkds I asked @erwango to take a look nad try it on his stm32f3_disco board. I don't think he has time this week. Linaro Connect HKG18 takes place this week so Linaro's team is busy.

**Comment by aakkds** [comment]
> @aakkds I asked @erwango to take a look and try it on his stm32f3_disco board. I don't think he has time this week. Linaro Connect HKG18 takes place this week so Linaro's team is busy.

@ydamigos Ok, thanks for the info!

**Comment by erwango** [comment]
@avisconti , would be nice if you could have a look

**Comment by avisconti** [comment]
@erwango @ydamigos @aakkds @MaureenHelm 
Guys, I was wondering whether this is the right approach or not.
Let me be frank and open: we do not need to rewrite evry time a new driver for sensors that basically
are the same one, or at least the same register space. For example, the lsm303dlhc accel and lis3dh accel can be handled by same driver (who_am_i is 0x33).
This is quite clear in linux where there is a common driver for all the sensors divided by who_am_i 
section. 

So, I'm not sure exactly how to organize this in the driver section, but it looks to me that it
should be possible to name the driver correctly and re-use them among different sensor thru a
compatibility list. For example, in case of lis3dh and lsm303dlhc accel, we may use something
like st_acc33 (0x33 is the wai).

The list of sensor is long, but I'm not 100% sure it is worth the hassle, so I just want to hear your
opinion. What do you think?

**Comment by avisconti** [comment]
>  I agree with you. I was checking the linux implementation for l3gd20 gyroscope and find out that a 
> common driver is used for many ST gyroscopes. We should follow the same approach in Zephyr to 
> avoid duplication of code and help with maintenance.
@ydamigos 
Correct.
Basically there are two types of drivers that we are implementing in linux.
1. Drivers for the 'simple' sensors.
    Ex:
    drivers/iio/accel/st_accel_core.c
    drivers/iio/gyro/st_gyro_core.c
    drivers/iio/magnetometer/st_magn_core.c

     Inside these drivers there is a division by WAI for common defs.

2. Drivers for 'complex' sensors

     The  new sensors are getting more and more complex, with big fifos, emebedded sensor
     hub, i3c support and so on, like (lsm6ds3, lsm6ds3h, lsm6dsl, lsm6dsm, ...) For this sensor we 
      decided to go in linux with separate drivers. Still, they are sometimes reused (ex: lsm6dsm driver is
     declared compatible with lsm6dsl).

I think we try to do the second option always.
For example, in this case if we try to use the lis3dh driver for the lsm303dlhc accel, I'm
expecting that works. It should. So we may declare it compatible in the same way linux
does.

**Comment by avisconti** [comment]
@erwango 
Can we use the compatible string inside the drivers? I don't see any drivers using it.
But this will help us in re-using same driver among different sensors.

**Comment by erwango** [comment]
> Can we use the compatible string inside the drivers? I don't see any drivers using it.

It depends on how you want to use it. Until now, to fulfil small footprint requirement, this was not done
since it requires some dedicated code to know which driver you're using. Ideal use would be to be able to decide which piece of driver is needed before compilation.

Anyway, I think this is a pretty  good topic, but I propose to discuss it in a dedicated issue, and keep current PR going on. Created #7001 (so close) for that matter.

**Comment by ydamigos** [comment]
@avisconti Could you also check the magnetic sensor driver? On stm32f3_disco, I have problem with magnetic sensor measurements. After reset, sometimes it works, some it doesn't (the value doesn't update).

**Comment by avisconti** [comment]
> Could you also check the magnetic sensor driver in the PR? On stm32f3_disco, I have problem 
> with magnetic sensor measurements. After reset, sometimes it works, sometimes it doesn't 
> (the value doesn't update).

@ydamigos 
I may try to reproduce the issue maybe tomorrow.
Can you give me the sensor configuration? (fullscale, odr, ...)

File: drivers/sensor/lsm303dlhc_accel/Kconfig
Comment (avisconti) [reply drivers/sensor/lsm303dlhc_accel/Kconfig]: I had similar comment as well. All the sensor drivers are done in the same way, but I do not like it personally.
Maybe you can define it as 'int' with a 'range' between 0 and 9, explaining what number is what freq.

File: drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.h
Comment (avisconti) [inline drivers/sensor/lsm303dlhc_accel/lsm303dlhc_accel.h]: I think you should multiply by SENSOR_G (like the lis3dh does) to obtain m/s2.
Now I see values in G when I use this driver (as I already commented I'd be in favour to re-use the lis3dh driver for the
accelerometer of the 303dlhc, as they are basically the same sensor).

File: drivers/sensor/CMakeLists.txt
Comment (avisconti) [inline drivers/sensor/CMakeLists.txt]: Usually the short word 'magn' is used to indicate magnetometer and not 'magneto'.
I suggest to use always 'magn'. If you agree to this rule, pls apply it everywhere in the commit.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (avisconti) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: Is the above sequence correct?
Looking at the datasheet seems that magneto_buf[0] is the LSB
and magneto_buf[1] is the MSB. And so on. So you are populating
the samples in the wrong order. Does it make sense to you?

**Comment by avisconti** [comment]
@ydamigos 
I took the commits and I'm trying on my h/w setup (nucleo f411re + sensor shield) and
I have issue with the mag (it seems that I always read 0x0 all the regs).
The accel part works fine but in my opinion we need to converge to a single driver for that
particular accel sensor (0x33).

I reviewed the mag commit and I think there are few things to be changed. The most important two
are:

 1. The samples seems to be wronlgy interpreted (LSB with MSB for each axis)
 2. Polling the DRDY 10 time with a 100ms delay may not work if the odr is very
     low (like 1hz or 0.75hz).

Pls let me know if you have time to take a look at it.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (avisconti) [inline drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: I think that a fixed sleep to 100ms does not work.
If you have set the odr to <= 1hz you may always find DRDY to 0.
If you have set the odr to high freq you wait too long.

I think that a variable time based on odr would be suit better.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: I think you looked at the acceleration register. In the case of the acceleration, the LSB is at lower address, but for the magnetometer the MSB comes first (cf table 17)
Accel: OUT_X_L_A then OUT_X_M_A
Magn: OUT_X_H_M then OUT_X_L_M

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: I agree for the variable time

**Comment by aakkds** [comment]
@avisconti Also I am not familiar with the linux kernel, it sure looks it would be nice to share code between sensors but as @erwango this would probably for another PR.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (avisconti) [reply drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: Correct, sorry! I have probably mis-looked at the datasheet.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (avisconti) [reply drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: ok, and if you do in this case 10 retries (10 ODRs basically) are too much now.
You may want to decrease to 3/4. What do you think?

**Comment by avisconti** [comment]
@aakkds 
There was also another comment of mine. 
Right now it seems that you are returning accel samples in G, while I think
you should report them in m/s2. So, as lis3dh does, you need to multiply by
SENSOR_G.

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: Mh I thought about keeping the 10 retries but having the sleep time to be ODR/10, or maybe even 11 retries if the querry is perfectly synchronised with the ODR.

**Comment by aakkds** [comment]
@avisconti Yep sorry I missed this comment and the range in Kconfig, I ll look at that this week end!

File: drivers/sensor/lsm303dlhc_accel/Kconfig
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_accel/Kconfig]: Yes this seems good to me! Sorry @MaureenHelm , by re reading this, I don't know why I didn t get what you were saying the first time :(

File: drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c
Comment (aakkds) [reply drivers/sensor/lsm303dlhc_magneto/lsm303dlhc_magneto.c]: Oh I wasn't aware of this, thanks! It will be much more easier to break the 80 char line rule :)

**Comment by aakkds** [comment]
@ydamigos 
Argh, I m actually considering buying a stm32f3_disco to see what is going on :(

**Comment by ydamigos** [comment]
>Argh, I m actually considering buying a stm32f3_disco to see what is going on :(

My stm32f3_disco board is rev. B. I don't know if the same problem exists in rev. C boards or if you could find rev. B in the market.

**Comment by avisconti** [comment]
@ydamigos 
Instead of adding a k_sleep(1) there can you possibly try to set the odr as last step instead of first step?
Maybe it won't fix the issue but I think this is the way it should always be.

**Comment by ydamigos** [comment]
>Instead of adding a k_sleep(1) there can you possibly try to set the odr as last step instead of first step?
Maybe it won't fix the issue but I think this is the way it should always be.

@avisconti I tried it and the issue remains. I tried also `k_sleep(1)` in init function before configuring the sensor and it works. I also noticed that the problem exists when I apply power to the board not during a reset using reset button. If I apply power and reset it three times it works again. Is it possible the magnetic sensor to need some time for initialization during power up?

**Comment by aakkds** [comment]
@avisconti @tbursztyka I addressed your comments.

However, I completely broke shippable, I'll fix that tomorrow/wednesday

**Comment by ydamigos** [comment]
>I have issue with the mag (it seems that I always read 0x0 all the regs).

@avisconti If you introduce some delay using k_sleep during sensor init, do you read meaningful values on your h/w setup (nucleo f411re + sensor shield)?

**Comment by erwango** [comment]
@aakkds , @avisconti , @ydamigos : What is the status on this PR?

**Comment by ydamigos** [comment]
@erwango If I introduce some delay using k_sleep during sensor init, I read meaningful magnetic sensor measurements on stm32f3_disco board, when I apply power on the board (that's when the issue exists). 

If others don't face any issue, I will unblock the PR.

**Comment by avisconti** [comment]
@ydamigos 
I completely forgot about this PR. I can summarize it in the following points:
  1. We require to re-organize all sensor drivers in such a way the code is reused. For this a new
      PR (#7001) was introduced. I may work on it if required in the next future.
   2.  On the mag driver seems that a delay is required in order to make it working

So, I'm not sure on what to do next. One possibility would be to add the k_sleep and merge the PR. In the future we may want to re-organize all the drivers.

On the other hand we may want to start the restructuring right now, which means that instead of merging this driver we will do the proper changes to re-use other drivers (0x33 is common between many sensors). Maybe @aakkds can do that if he wants to add contributions to zephyr.

I do not want to block anyone work, but I would prefer to have a clean way to work in the sensor arena.
Any comment?

**Comment by aakkds** [comment]
@ydamigos @avisconti 
I can add the delay in the init, @ydamigos was there a specific delay needed or 1ms is enough?
Concerning the refactoring of the sensors area, I think it would make more sense to do it from a dedicated branch. 
If you think it's better to not merge this driver and do the refactoring right away, I have no problems if you take care of it @avisconti, as I don't have that many sensors to test the changes.

**Comment by ydamigos** [comment]
>@ydamigos was there a specific delay needed or 1ms is enough?

@aakkds I tried it with 1ms delay.

**Comment by avisconti** [comment]
@aakkds 
We can target this for next release (i.e. 1.13).
The LSM303DLHC is a very old device (10 years old I guess) which has following sensors inside:
   - accelemoter (WAI 0x33). Same as lis3dh and lis2dh. Driver may be reused for all 3 accel
   - magnetometer. We can use the mag part of this driver

In linux sharing a driver is very simple, as you may add multiple "compatible" string in the driver. Maybe for zephyr is the same. So, maybe you can push this driver and I will restructure it later on.

Any other opinion?

**Comment by erwango** [comment]
> In linux sharing a driver is very simple, as you may add multiple "compatible" string in the driver. Maybe for zephyr is the same. 

We target similar dts support in Zephyr than in Linux. So far using multiple "compatible" is working in some cases, but I can't say it'll work for all cases. If not the case we'll fix it

**Comment by aakkds** [comment]
@avisconti 
Ok, I ll do that (probably won't be home until sunday).
Thanks!

**Comment by galak** [comment]
Any update on this driver, its been a while since this PR seems to have been touched?  Should we close this out?

**Comment by aakkds** [comment]
@galak Sorry for the late reply. I think @avisconti didn't want to merge this directly but later, as part of a refactoring of the sensors drivers.
Other than that, I made the requested changes

**Comment by nashif** [comment]
should we close this in favor of #6441?

**Comment by MaureenHelm** [comment]
Superseded by #6441

**Pull request closed without merging**
